File: 2022m./uzd2./sprendimai/uzd2.py
class Exercise:
    daysDone = 0
    morningsDone = 0
    middaysDone = 0
    eveningsDone = 0
    timeSpent = 0
    def __init__(self, name):
        self.name = name

def findAllExercises():
    data = open("U2.txt", "r")

    exercises = [None]*5
    exercisesFound = 0

    for i in range(int(next(data))):
        dayData = next(data)
        while("  " in dayData):
            dayData = dayData.replace("  ", " ")
        dayData = dayData.split(" ")
        for j in range(int(dayData[0])):
            if dayData[1 + j * 3] not in exercises:
                exercises[exercisesFound] = dayData[1 + j * 3]
                exercisesFound += 1
                if(exercisesFound == 5):
                    break
        if(exercisesFound == 5):
            break

    data.close()

    return exercises

def main():
    data = open("U2.txt", "r")
    output = open("U2rez.txt", "w")

    exerciseNames = findAllExercises()
    exercises = [Exercise(exerciseNames[0]), Exercise(exerciseNames[1]), Exercise(exerciseNames[2]), Exercise(exerciseNames[3]), Exercise(exerciseNames[4])]

    for i in range(int(next(data))):
        dayData = next(data)
        while("  " in dayData):
            dayData = dayData.replace("  ", " ")
        dayData = dayData.split(" ")
        for exercise in exercises:
            if(exercise.name in dayData):
                exercise.daysDone += 1
        for j in range(int(dayData[0])):
            exerciseIndex = exerciseNames.index(dayData[1 + j * 3])
            if(dayData[2 + j * 3] == "Rytas"):
                exercises[exerciseIndex].morningsDone += 1
            elif(dayData[2 + j * 3] == "Diena"):
                exercises[exerciseIndex].middaysDone += 1
            else:
                exercises[exerciseIndex].eveningsDone += 1
            exercises[exerciseIndex].timeSpent += int(dayData[3 + j * 3])

    exercises.sort(key=lambda x: x.name)

    for exercise in exercises:
        filler = " "*(14-len(exercise.name))
        output.write(f"{exercise.name}{filler} {exercise.daysDone} {exercise.timeSpent}\n")
        if(exercise.morningsDone > 0): output.write(f"Rytas   {exercise.morningsDone}\n")
        if(exercise.middaysDone > 0):  output.write(f"Diena   {exercise.middaysDone}\n")
        if(exercise.eveningsDone > 0): output.write(f"Vakaras {exercise.eveningsDone}\n")

    data.close()
    output.close()

if __name__ == "__main__":
    main()

File: 2022m./uzd2./sprendimai/uzd2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Exercise {
public:
    int daysDone = 0;
    bool doneToday = false;
    int morningsDone = 0;
    int middaysDone = 0;
    int eveningsDone = 0;
    int timeSpent = 0;
    string name;

    Exercise(const string& name) : name(name) {}
};

vector<string> findAllExercises() {
    ifstream data("U2.txt");

    vector<string> exercises(5);
    int exercisesFound = 0;
    int numEntries;
    data >> numEntries;

    while (exercisesFound < 5 && numEntries--) {
        int numExercises;
        data >> numExercises;

        for (int j = 0; j < numExercises; j++) {
            if (exercisesFound == 5) {
                break;
            }

            string exercise, extra, extra2;
            data >> exercise >> extra >> extra2;

            if (find(exercises.begin(), exercises.end(), exercise) == exercises.end()) {
                exercises[exercisesFound] = exercise;
                exercisesFound++;
            }
        }
        data.ignore();
    }

    data.close();

    return exercises;
}

int main() {
    ifstream data("U2.txt");
    ofstream output("U2rez.txt");

    vector<string> exerciseNames = findAllExercises();
    vector<Exercise> exercises;

    for (const string& exerciseName : exerciseNames) {
        exercises.push_back(Exercise(exerciseName));
    }

    int days;
    data >> days;

    while (days > 0) {
        int dayExercises;
        data >> dayExercises;
        for (int i = 0; i < dayExercises; ++i) {
            string exerciseName, timeOfDay;
            int timeSpent;
            data >> exerciseName >> timeOfDay >> timeSpent;

            for (Exercise& exercise : exercises) {
                if (exercise.name == exerciseName) {
                    if (!exercise.doneToday) {
                        exercise.daysDone++;
                        exercise.doneToday = true;
                    }
                    exercise.timeSpent += timeSpent;
                    if (timeOfDay == "Rytas") {
                        exercise.morningsDone++;
                    } else if (timeOfDay  == "Diena") {
                        exercise.middaysDone++;
                    } else {
                        exercise.eveningsDone++;
                    }
                    break;
                }
            }
        }
        for (Exercise& exercise : exercises) {
            exercise.doneToday = false;
        }
        days--;
    }

    sort(exercises.begin(), exercises.end(), [](const Exercise& a, const Exercise& b) {
        return a.name < b.name;
    });

    for (const Exercise& exercise : exercises) {
        string filler = "               ";
        filler.erase(14 - exercise.name.length());
        output << exercise.name << filler << " " << exercise.daysDone << " " << exercise.timeSpent << "\n";
        if (exercise.morningsDone > 0) output << "Rytas   " << exercise.morningsDone << "\n";
        if (exercise.middaysDone > 0)  output << "Diena   " << exercise.middaysDone << "\n";
        if (exercise.eveningsDone > 0) output << "Vakaras " << exercise.eveningsDone << "\n";
    }

    data.close();
    output.close();

    return 0;
}